File: visit.py
import requests
import dataset
import arrow
import csv
import asyncio
import os
import pyppeteer


# config
initial_date = '2018-01-01'

# setup db
db = dataset.connect("sqlite:///data/webcat.db")

async def get_screenshots(site, date_short):

    defaultSizes = {
        "0600x0": {"width": 600, "height": 0},
        "1200x0": {"width": 1200, "height": 0}}

    # make sure path exists
    screenshot_path = f'data/screenshots/{site["slug"]}/{date_short}/'
    os.makedirs(screenshot_path, exist_ok=True)

    for key, value in defaultSizes.items():

        print(key, value)

        # # setup
        # browser = await pyppeteer.launch(
        #     autoClose=False
        # )
        # page = await browser.newPage()
        # url = f'https://web.archive.org/web/{date_short}/{site["url"]}'
        # await page.goto(url, {
        #     'waitUntil': 'networkidle0'
        # })

        # # remove wayback crap
        # await page.evaluate('''() => {
        #     let dom = document.querySelector('#wm-ipp');
        #     dom.parentNode.removeChild(dom);
        # }''')

        # # take screenshot
        # await page.setViewport(value)
        # await page.screenshot({
        #     'path': f'{screenshot_path}/{key}.png',
        #     'fullPage': True
        # })

        # # cleanup
        # await page.close()
        # await browser.close()



# async def scrape_wayback(browser, site, date_input):
async def scrape_wayback(site, date_input):
    # get correct date
    wb_format = "YYYYMMDDHHmmss"
    date_wanted = date_input.format(wb_format)
    date = arrow.get(
        requests.get(
            f'http://archive.org/wayback/available?url={site["url"]}&timestamp={date_wanted}'
        ).json()["archived_snapshots"]["closest"]["timestamp"],
        wb_format,
    )
    date_short = date.format('YYYY-MM-DD')

    # see if scrape already exists in db
    table = db['scrapes']
    exists = table.find_one(date_scrape=date.to('utc').isoformat())

    if not exists:
        date_wb = date.format(wb_format)
        wayback_url = f'http://web.archive.org/web/{date_wb}id_/{site["url"]}'

        # get site
        r = requests.get(wayback_url)

        # store it
        table = db['scrapes']
        table.insert({
            'slug': site['slug'],
            'url_site': site['url'],
            'url_wayback': wayback_url,
            'response_content_type': r.headers['content-type'],
            'response_encoding': r.encoding,
            'response_text_utf8': r.text,
            'date_wayback': date.to('utc').isoformat(), 
            'date_retrieved': arrow.now().isoformat()
        })

        # get screenshot
        await get_screenshots(site, date_wb)

        # report to screen
        print(f'OK: {site["slug"]} {date_short}')
    
    else:
        print(f'--: {site["slug"]} {date_short} (exists)')


async def main():

    # setup site
    with open("data/libraries.csv") as libraries_csv:
        reader = csv.reader(libraries_csv)

        cols = next(reader)
        for row in reader:
            # setup date
            date = arrow.get(initial_date)

            # setup site
            site = {}
            for col_index, col in enumerate(row):
                col_name = cols[col_index]
                site[col_name] = col
            
            # iterate sites/dates
            while date <= arrow.utcnow():
                await scrape_wayback(site, date)
                date = date.shift(years=+1)


if __name__ == "__main__":
    asyncio.run(main())

File: visit.js
const request = require("superagent");
const fs = require("fs");
const moment = require("moment");
const csvParse = require("csv-parse/lib/sync");
const colors = require("colors");
const puppeteer = require("puppeteer");
// const rm = require("rimraf");
const startDate = "1995-01-01";


// main
(async () => {
  try {
    // setup
    const browser = await puppeteer.launch();

    // parse libraries
    const librariesCSV = await fs.promises.readFile(
      "data/libraries.csv",
      "utf8"
    );
    const libraries = csvParse(librariesCSV, {
      columns: true,
      skip_empty_lines: true
    });

    // iterate libraries
    for (const library of libraries) {
      // iterate dates
      let date = moment.utc(startDate);
      while (date.isBefore(moment())) {
        // scrape
        await scrapeWayback(browser, library, date);
        date = date.add(1, "years");
      }
    }

    // cleanup
    await browser.close();
  } catch (error) {
    console.error(error);
  }
})();

async function scrapeWayback(browser, library, dateInput) {
  const wbFormat = "YYYYMMDDHHmmss";

  let date;

  try {
    const r = await request.get(
      `https://archive.org/wayback/available?url=${
        library.url
      }/&timestamp=${dateInput.format(wbFormat)}`
    );
    date = moment.utc(r.body.archived_snapshots.closest.timestamp, wbFormat);
    const waybackUrl = `http://web.archive.org/web/${date.format(wbFormat)}/${
      library.url
    }`;
    const waybackRawUrl = `http://web.archive.org/web/${date.format(
      wbFormat
    )}id_/${library.url}`;

    const waybackRaw = await request.get(waybackRawUrl);

    let output = {
      slug: library.slug,
      site: library.url,
      date: dateInput.format(),
      dateRetrieved: moment.utc().format(),

      // for data from puppeteer, for the most part
      rendered: {
        url: waybackUrl,
        page: {},
        browser: {}
      },

      // for raw HTML, from Superagent
      raw: {
        url: waybackRawUrl,
        response: {
          status: waybackRaw.status,
          type: waybackRaw.type,
          headers: waybackRaw.header, 
        }
      }
    };

    // filesystem
    const metaDir = `data/scrapes/meta`;
    const pagesDir = `data/scrapes/pages`;
    const screensDir = `data/scrapes/screens`;
    await fs.promises.mkdir(metaDir, { recursive: true });
    await fs.promises.mkdir(pagesDir, { recursive: true });
    await fs.promises.mkdir(screensDir, { recursive: true });

    // puppeteer
    try {
      const viewports = {
        // "0001x0001": {
        //   width: 1,
        //   height: 1,
        //   // isMobile: false,
        //   isLandscape: false
        // },
        "0600": {
          width: 600,
          height: 1
          // isMobile: true,
          // isLandscape: false
        },
        "1200": {
          width: 1200,
          height: 1
          // isMobile: false,
          // isLandscape: true
        }
      };

      // scraping
      for (const v in viewports) {
        const page = await browser.newPage();

        // Navigate to page
        await page.goto(waybackUrl, {
          waitUntil: "networkidle0"
          // timeout: 100000
        });

        // strip wayback div
        await page.evaluate(() => {
          let waybackDiv = document.querySelector("#wm-ipp");
          waybackDiv.parentNode.removeChild(waybackDiv);
        });
        await page.setViewport(viewports[v]);

        // screenshots
        const screenFile = `${screensDir}/${library.slug}-${date.format(wbFormat)}-${v}.png`;
        if (!fs.existsSync(screenFile)) {
          await page.screenshot({
            path: screenFile,
            fullPage: true,
            omitBackground: true,
          });
          console.log(
            `${colors.green("scrn OK")}: ${library.slug}-${date.format(
              wbFormat
            )}-${v}.png`)
        } else {
          console.log(
            `${colors.yellow("scrn --")}: ${library.slug}-${date.format(
              wbFormat
            )}-${v}.png`
          );
        }

        // collect extra puppeteer metadata
        output.rendered.page.metrics = await page.metrics();
        output.rendered.page.title = await page.title()
        // output.wayback.page.accessibility = await page.accessibility.snapshot();
        output.rendered.browser.userAgent = await browser.userAgent();
        output.rendered.browser.version = await browser.version();

        // cleanup
        await page.close();
      }
      // );
    } catch (error) {
      console.log(
        // `${colors.red("NOPE   ")}: ${date.format(wbFormat)} ${
        //   library.slug
        // } (${colors.red(error.name)})`
        error
      );
    }

    // output metadata
    const mdFile = `${metaDir}/${library.slug}-${date.format(wbFormat)}.json`;
    if (!fs.existsSync(mdFile)) {
      await fs.promises.writeFile(mdFile, JSON.stringify(output));
      console.log(
        `${colors.green("meta OK")}: ${library.slug}-${date.format(
          wbFormat
        )}.json`
      );
    } else {
      console.log(
        `${colors.yellow("meta --")}: ${library.slug}-${date.format(
          wbFormat
        )}.json`
      );
    }

    // output raw html
    const pageFile = `${pagesDir}/${library.slug}-${date.format(wbFormat)}.html`;
    if (!fs.existsSync(pageFile)) {
      await fs.promises.writeFile(pageFile, waybackRaw.text);
      console.log(
        `${colors.green("page OK")}: ${library.slug}-${date.format(
          wbFormat
        )}.html`
      );
    } else {
      console.log(
        `${colors.yellow("page --")}: ${library.slug}-${date.format(
          wbFormat
        )}.html`
      );
    }


  } catch (error) {
    console.error(error);
  }
  // const result = await axios.get(library.url);
  // console.log(result.data)
}
